src/clients/AcrossAPIClient.ts

```
@@ -0,0 +1,80 @@
+import { winston, BigNumber, MAX_UINT_VAL } from "../utils";
+import { l2TokensToL1TokenValidation } from "../common";
+import axios, { AxiosError } from "axios";
+import get from "lodash.get";
+
+export interface DepositLimits {
+  maxDeposit: BigNumber;
+}
+
+export class AcrossApiClient {
+  private endpoint = "https://across.to/api/";
+
+  private limits: { [token: string]: BigNumber } = {};
+
+  constructor(readonly logger: winston.Logger, readonly tokensQuery: string[] = [], readonly timeout: number = 5000) {
+    if (Object.keys(tokensQuery).length === 0) this.tokensQuery = Object.keys(l2TokensToL1TokenValidation);
+  }
+
+  async update(): Promise<void> {
+    this.logger.debug({
+      at: "AcrossAPIClient",
+      message: "Updating AcrossAPI client",
+      timeout: this.timeout,
+      tokensQuery: this.tokensQuery,
+      endpoint: this.endpoint,
+    });
+
+    // /limits
+    // - Store the max deposit limit for each L1 token. DestinationChainId doesn't matter since HubPool
+    // liquidity is shared for all tokens and affects maxDeposit. We don't care about maxDepositInstant
+    // when deciding whether a relay will be refunded.
+    const data = await Promise.all(
+      this.tokensQuery.map((l1Token) => {
+        const validDestinationChainForL1Token = l2TokensToL1TokenValidation[l1Token]
+          ? Number(Object.keys(l2TokensToL1TokenValidation[l1Token])[0])
+          : 10;
+        return this.callLimits(l1Token, validDestinationChainForL1Token);
+      })
+    );
+    for (let i = 0; i < this.tokensQuery.length; i++) {
+      const l1Token = this.tokensQuery[i];
+      this.limits[l1Token] = data[i]?.maxDeposit ? data[i].maxDeposit : BigNumber.from(MAX_UINT_VAL);
```

Curious about two points here:
 - Do we want to fail open or fail closed on a failed lookup?
 - If we want to fail open, MAX_UINT_VAL risks being greater than LP deposits.

On this basis, I'd be inclined to fail closed and default to 0. This means we'd be unable to make fills in the event of an API outage, but protects us against accepting deposits that are too much for us to really handle.

If we fail open, then we might instead want to default to a configurable value from the environment (...although we should be wary of adding even more configuration items, because they become hard to manage...).

wdyt?

---

Failing close is risky though. For example yesterday the API was hammered hard and was overloaded. This means the reliability of the bots would be affected by the performance of the API, which increases risks of things going down together like dominoes. As long as we log the errors as at least warnings, we can be alerted and react to this situation.

---

hmmm this is a great point @pxrl. I'm inclined also to fail close and default to 0, thus blocking fills. We should `warn` when the API fails, as @kevinuma suggests, and then what if we add an easy flag like `useLimits` that we can use to easily bypass the API incase its down for a known reason

---

I've thought about it some more, and I think it's preferable to fail closed, rather than open.

In the scenario where the API is down and we can't query limits, the FE and aggregators are going to be down too, because they probably can't query suggested-fees or limits anyway. This could even be dangerous if they mask the errors and proceed with default values that we might never have seen in actual deposits.

At the same time, if API endpoints are down then we should expect that the vast majority of new deposits reduces dramatically or entirely. Therefore, I think taking marginal risks to ensure the bots keep operating seems to have a marginal benefit at best. In the worst case, we would be able to make local changes to the bot to flush and outstanding deposits.

A possible compromise would be to disable slow fills in the event that we can't query /limits. This puts a pretty low ceiling on the amount of possible damage that could be done.

---

I agree with failing closed and I've implemented that